---
output:
  html_document:
    toc: true
    toc_float: false
  pdf_document:
    toc: true
params:
  rdata_path:
    value: ""
---

```{r setup, include=FALSE}
library(tidyverse)
library(flextable)

# Avoid scientific notation
options(scipen = 999)
# Configure Rmarkdown/knitr/pandoc processing
knitr::opts_chunk$set(echo = FALSE,
                      fig.align = "center",
                      out.width = "80%")
# Global ggplot2 theme
theme_set(theme_minimal())
# Table style and theme
set_flextable_defaults(
  font.family = "Verdana",
  padding = 5,
  tabcolsep = 5,
  table.layout = "autofit",
  theme_fun = "theme_vanilla"
)

# Load precomputed dataframes etc.
load(params$rdata_path)
```

---
title: "AAV Sequence Analysis"
subtitle: 'Sample: `r r_params$sample_id`'
date: 'Date: `r format(Sys.Date())`'
---


`\newpage`{=latex}

# Read type classification

## Definitions

Assigned Type|Definition
------------:|----------------------------------------------------
scAAV        |Self-complementary AAV where one half of the payload region is a reverse complement of the other, resulting in an intra-molecular double-stranded DNA template. A sequencing read is inferred as scAAV if it has both a primary and supplementary alignment to the vector genome.
ssAAV        |AAV where the resulting DNA template is expected to be single-stranded, as opposed to self-complementary. A sequencing read is inferred as ssAAV if it has a single alignment to the AAV genome and no complementary secondary alignment.
other        |Read consists of a fragment mapping to the vector but with unexpected polarities (e.g. +,-,- or +,+,+) and cannot be well-defined at the moment. This means that the algorithm was not able to distinguish the read either as ssAAV or scAAV by the definitions above, usually due to multiple supplementary alignments on the vector region or other unusual alignment features.
host         |Read originates from the host genome that is given (e.g. hg38, CHM13).
repcap       |Read originates from the repcap plasmid. The Rep gene encodes four proteins (Rep78, Rep68, Rep52, and Rep40), which are required for viral genome replication and packaging, while Cap expression gives rise to the viral capsid proteins (VP; VP1/VP2/VP3), which form the outer capsid shell that protects the viral genome, as well as being actively involved in cell binding and internalization.
helper       |Read originates from the helper plasmid. In addition to Rep and Cap, AAV requires a helper plasmid containing genes from adenovirus. These genes (E4, E2a and VA) mediate AAV replication.
chimeric     |Read consists of fragments that map to one or more "genomes" (e.g. vector and host; helper and repcap).

*Note:* Even though ssAAV distinguishes one ITR as the wildtype (wtITR) and the other as the mutated ITR (mITR), we will still refer to them as "left ITR" and "right ITR". For example, "left-partial" would be equivalent to "mITR-partial" in the case where the mITR is the left ITR based on the given genomic coordinates.

`<br />`{=html}

![Single-stranded AAV types (ssAAV).
ssAAV reads must fully map within the vector genome (which can contain the backbone, beyond the ITR region) consisting of a single (primary) alignment.
"ssAAV-full" reads must cover from left ITR to right ITR.
"ssAAV-left-partial" contain the left ITR but are missing the right ITR.
"scAAV-vector+backbone" map partially within the ITR and partially to the backbone.
](class_def_ssAAV.png)

`<br />`{=html}

![Self-complementary AAV types (scAAV).
scAAV reads must fully map within the vector genome (which can contain the backbone, beyond the ITR region) consisting of both primary and supplementary alignments.
The definitions of full, partial, and backbone reads are the same as in ssAAV.
](class_def_scAAV.png)

`\break`{=latex}

# Assigned types by read alignment characteristics

```{r sopt1, message=FALSE, warning=FALSE}
flextable(df.read1) %>%
  set_header_labels(values = c("Assigned Type", "Count", "Frequency (%)"))
```

```{r barpercentreads, message=FALSE, warning=FALSE}
countreads <- df.read1 %>% dplyr::filter(assigned_type != 'Lambda')

# Bar chart
ggplot(countreads, aes(x = assigned_type, y = freq, fill = assigned_type)) +
  geom_bar(stat = 'identity') +
  xlab("Sequence") +
  scale_y_continuous("Reads (%)",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

# Pie chart
ggplot(countreads, aes(x = "", y = e_count, fill = assigned_type)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar("y", start = 0) +
  theme_void()
```


## Single-stranded vs self-complementary frequency

```{r sopt2, message=FALSE, warning=FALSE}
sopt2 <- df.read.vector1 %>%
  mutate(totfreq = round(100 * e_count / total_read_count.all, 2))

flextable(sopt2) %>%
  set_header_labels(values = c(
    "Assigned Type",
    "Count",
    "Frequency in AAV (%)",
    "Total Frequency (%)"
  ))
```

# Distribution of read lengths by assigned AAV types

```{r atypeviolin, message=FALSE, warning=FALSE}
allowed_subtypes <- c('full', 'full-gap', 'vector+backbone')

p2.atype_violin <- ggplot(
  filter(x.read.vector, assigned_subtype %in% allowed_subtypes),
  aes(
    x = paste(assigned_type, assigned_subtype, sep = '-'),
    y = read_len
  )
) +
  geom_violin() +
  labs(title = "Distribution of read lengths by assigned AAV type",
       x = "Assigned AAV type") +
  scale_y_continuous("Read length",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0))

p2.atype_violin
```

# Assigned AAV read types detailed analysis

## Assigned AAV types (top 20)

```{r sopt3, message=FALSE, warning=FALSE}
sopt3 <- df.read.vector2 %>%
  mutate(totfreq = round(100 * e_count / total_read_count.all, 2)) %>%
  arrange(
    factor(assigned_type, levels = valid_types),
    factor(assigned_subtype, levels = valid_subtypes)
  )

flextable(sopt3[1:min(nrow(sopt3), 20), ]) %>%
  set_header_labels(
    values = c(
      "Assigned Type",
      "Assigned Subtype",
      "Count",
      "Freq. in AAV (%)",
      "Total Freq. (%)"
    )
  )
```

## Definitions

Assigned Subtype|Definition
---------------:|----------------------------------------------------------------
full            |Read alignment includes the entire ITR-to-ITR target vector sequence.
left-partial    |Read aligns to a fragment of the vector originating from the left (upstream) ITR of the vector while not covering the right ITR.
right-partial   |Read aligns to a fragment of the vector originating from the right (downstream) ITR of the vector while not covering the left ITR.
partial         |Read aligns to a fragment of the vector originating from within the ITR sequences.
vector+backbone |Read aligns to a fragment including the vector as well as plasmid backbone sequence. May imply read-through beyond the right ITR, or reverse packaging if the alignment is to only the left ITR and backbone.
backbone        |Read aligns to a fragment originating solely from the plasmid backbone sequence.
snapback        |Read consists of a double-stranded, sub-genomic fragment including only one ITR and read alignments in both (+) and (-) polarities. (ssAAV only)


# Flip/flop considerations

Term|Definition
---:|----------------
Flip/Flop|One ITR is formed by two palindromic arms, called B–B' and C–C', embedded in a larger one, A–A'. The order of these palindromic sequences defines the flip or flop orientation of the ITR. ([Read more](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5655423/))


```{r flipflopplot, message=FALSE, warning=FALSE, results='asis'}
if (exists("df.flipflop")) {
  # Generate flip/flop summary plot
  ffplot <- df.flipflop %>%
    mutate(
      class = if_else(
        leftITR == 'unclassified' |
          rightITR == 'unclassified' |
          leftITR == 'unknown' | rightITR == 'unknown',
        "unclassifed",
        paste(leftITR, rightITR, sep = '-')
      )
    ) %>%
    group_by(type, subtype, class) %>%
    summarize(classct = sum(count)) %>%
    filter(subtype == 'vector-full')
  ggplot(ffplot, aes(x = type, y = classct, fill = class)) +
    geom_bar(stat = 'identity') +
    labs(x = 'Assigned type') +
    scale_y_continuous("Number of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

} else {
  cat("No flip/flop analysis results available to display.")
}
```

```{r flipflopsc, message=FALSE, warning=FALSE, results='asis'}
if (exists("scff")) {
  cat("## Flip/flop configurations, scAAV only\n\n")
  if (nrow(scff) > 1) {
    flextable(scff) %>%
      set_header_labels(values = c("type", "subtype", "leftITR", "rightITR", "count"))
  } else {
    cat("\nNo scAAV flip/flop analysis results available to display.")
  }
}
```

```{r flipflopss, message=FALSE, warning=FALSE, results='asis'}
if (exists("ssff")) {
  cat("## Flip/flop configurations, ssAAV only\n\n")
  if (nrow(ssff) > 1) {
    flextable(ssff) %>%
      set_header_labels(values = c("type", "subtype", "leftITR", "rightITR", "count"))
  } else {
    cat("No ssAAV flip/flop analysis results available to display.")
  }
}
```


# Distribution of read length by subtype

```{r lenhisto2_subtype, message=FALSE, warning=FALSE}
# scAAV length histogram
p1.scAAV_len_hist <- ggplot(filter(x.read.vector, assigned_type == 'scAAV'),
                            aes(x = read_len, color = subtype)) +
  geom_freqpoly() +
  labs(title = "Distribution of scAAV read length by subtype") +
  scale_x_continuous("Read length (bp)",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("Count",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

# ssAAV length histogram
p1.ssAAV_len_hist <- ggplot(filter(x.read.vector, assigned_type == 'ssAAV'),
                            aes(x = read_len, color = subtype)) +
  geom_freqpoly() +
  labs(title = "Distribution of ssAAV read length by subtype") +
  scale_x_continuous("Read length (bp)",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("Count",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p1.scAAV_len_hist
p1.ssAAV_len_hist
```


# AAV mapping to reference sequence

## Gene therapy construct

```{r lenhisto2_construct, message=FALSE, warning=FALSE}
p1.map_starts <- ggplot(x.summary.vector, aes(map_start0 + 1, fill = map_subtype)) +
  geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                 binwidth = 150,
                 boundary = 0,
                 closed = "left") +
  geom_vline(xintercept = TARGET_REGION_START,
             color = 'red',
             lty = 2) +
  geom_vline(xintercept = TARGET_REGION_END,
             color = 'red',
             lty = 2) +
  scale_x_continuous("Mapped reference start position",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("Fraction of reads",
                     limits = c(0, 1),
                     expand = c(0, 0)) +
  labs(title = "Distribution of mapped read starts on construct")

p1.map_ends <- ggplot(x.summary.vector, aes(map_end1, fill = map_subtype)) +
  geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                 binwidth = 150,
                 boundary = 0,
                 closed = "left") +
  geom_vline(xintercept = TARGET_REGION_START,
             color = 'red',
             lty = 2) +
  geom_vline(xintercept = TARGET_REGION_END,
             color = 'red',
             lty = 2) +
  scale_x_continuous("mapped reference end position",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("fraction of reads",
                     limits = c(0, 1),
                     expand = c(0, 0)) +
  labs(title = "Distribution of mapped read ends on construct")

p1.map_len <- ggplot(x.summary.vector, aes(map_len, fill = map_subtype)) +
  geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                 binwidth = 150,
                 boundary = 0,
                 closed = "left") +
  scale_x_continuous("mapped reference length",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("fraction of reads",
                     limits = c(0, 1),
                     expand = c(0, 0)) +
  labs(title = "Distribution of mapped spanning region sizes on construct")

p1.map_starts
p1.map_ends
p1.map_len
```

<!-- optional RepCap section -->

```{r lenhisto2_rc, message=FALSE, warning=FALSE, results='asis'}
x.read.repcap <- filter(x.all.read, assigned_type == 'repcap')

# Only render this subsection if enough reads mapped to repcap
if (dim(x.read.repcap)[1] > 10) {
  x.summary.repcap <- filter(x.all.summary, read_id %in% x.read.repcap$read_id)

  # Subsection heading in Markdown
  cat("## RepCap\n\n")

  p1.map_starts.repcap <- ggplot(x.summary.repcap, aes(map_start0 + 1)) +
    geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                   binwidth = 150,
                   boundary = 0,
                   closed = "left") +
    geom_vline(xintercept = TARGET_REGION_START_REPCAP,
               color = 'red',
               lty = 2) +
    labs(title = "Distribution of mapped read starts on RepCap") +
    scale_x_continuous("Mapped reference start position",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05))) +
    scale_y_continuous("Fraction of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

  p1.map_starts.repcap
}
```

```{r lenhisto2_rc_end, message=FALSE, warning=FALSE, results='asis'}
if (exists("x.summary.repcap")) {
  p1.map_ends.repcap <- ggplot(x.summary.repcap, aes(map_end1)) +
    geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                   binwidth = 150,
                   boundary = 0,
                   closed = "left") +
    geom_vline(xintercept = TARGET_REGION_END_REPCAP,
               color = 'red',
               lty = 2) +
    labs(title = "Distribution of mapped read ends on RepCap") +
    scale_x_continuous("Mapped reference end position",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05))) +
    scale_y_continuous("Fraction of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

  p1.map_ends.repcap
}
```

```{r lenhisto2_rc_len, message=FALSE, warning=FALSE, results='asis'}
if (exists("x.summary.repcap")) {
  p1.map_len.repcap <- ggplot(x.summary.repcap, aes(map_len)) +
    geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                   binwidth = 150,
                   boundary = 0,
                   closed = "left") +
    labs(title = "Distribution of mapped spanning region sizes on RepCap") +
    scale_x_continuous("Mapped reference length",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05))) +
    scale_y_continuous("Fraction of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

  p1.map_len.repcap
}
```


# Distribution of non-matches by reference position

```{r varplots, message=FALSE, warning=FALSE}
p1.err_sub <- ggplot(filter(df.err.vector, type == 'mismatch'),
                     aes(x = pos0_div, y = count)) +
  geom_bar(fill = 'darkgreen', stat = 'identity') +
  xlim(c(TARGET_REGION_START, TARGET_REGION_END)) +
  labs(title = "Substitutions by reference position",
       subtitle = "Higher bars indicate hot spots for substitutions w.r.t reference",
       x = "Reference position") +
  scale_y_continuous("Frequency",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p1.err_del <- ggplot(filter(df.err.vector, type == 'deletion'),
                     aes(x = pos0_div, y = count)) +
  geom_bar(fill = 'darkred', stat = 'identity') +
  xlim(c(TARGET_REGION_START, TARGET_REGION_END)) +
  labs(title = "Deletions by reference position",
       subtitle = "Higher bars indicate hot spots for deletion w.r.t reference",
       x = "Reference position") +
  scale_y_continuous("Frequency",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p1.err_ins <- ggplot(filter(df.err.vector, type == 'insertion'),
                     aes(x = pos0_div, y = count)) +
  geom_bar(fill = 'darkblue', stat = 'identity') +
  xlim(c(TARGET_REGION_START, TARGET_REGION_END)) +
  labs(title = "Insertions by reference position",
       subtitle = "Higher bars indicate hot spots for insertion w.r.t reference",
       x = "Reference position") +
  scale_y_continuous("Frequency",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p1.err_sub
p1.err_del
p1.err_ins
```


# Methods

This report was generated by an automated analysis of long-read sequencing data from
adeno-associated virus (AAV) products.
The sequencing data should be from the PacBio sequencer run in AAV mode, or equivalent
circular consensus sequencing (CCS) reads (Travers et al., 2010).

In this analysis, reads are aligned to the given AAV, packaging, and host reference
sequences using Minimap2 (Li, 2018).
The reference sequences for each primary alignment and its orientation are counted and
summarized to assign read type classifications, including vector, non-vector, and
chimeric reads.
For reads assigned to the AAV vector, the primary alignment coordinates are compared to
the annotated vector region in the reference sequence, which comprises the left and
right ITRs and the genomic region between them, to assign each read to a subtype
classification.
Sequence variants relative to the vector reference sequence are determined directly from
each read's alignment, specifically the CIGAR string indicating insertions, deletions,
mismatches, and gaps.


## Citations

1. Travers, K. J., Chin, C.-S., Rank, D. R., Eid, J. S. & Turner, S. W. A flexible and efficient template format for circular consensus sequencing and SNP detection. Nucleic Acids Research 38, e159–e159 (2010).
2. Li, H. Minimap2: Pairwise alignment for nucleotide sequences. Bioinformatics 34, 3094–3100 (2018).
